pkg/transforms/mqttsecret.go

```diff
@@ -192,6 +210,27 @@ func (sender *MQTTSecretSender) MQTTSend(ctx interfaces.AppFunctionContext, data
 	return true, nil
 }
 
+func (sender *MQTTSecretSender) formatTopic(ctx interfaces.AppFunctionContext, data interface{}) (string, error) {
+	publishTopic := sender.mqttConfig.Topic
+	var err error
+
+	if sender.topicFormatter != nil {
+		publishTopic, err = sender.topicFormatter(publishTopic, ctx, data)
+	} else {
+		for k, v := range ctx.GetAllValues() {
+			publishTopic = strings.Replace(publishTopic, fmt.Sprintf("{%s}", k), v, -1)
```

Comment:
Would using text/template make sense here instead of the replace?

Reply:
But this approach does have an advantage, since text/template wasn't working well with special key names such as those containing a "$" character.

Reply:
I thought it would be best to keep the complexities of text/template (like you mention) out of the default behavior.  It would require more complex expressions to render against the topic map as well, and wouldn't provide a way for us to catch "replacement not found" errors that I can see.  Have a look here:

https://play.golang.org/p/QwUSeJnLaTh

That said I did add the ability to specify your own topic format function as well, so if you wanted to do something with template that fits your use case it should be easy enough to plug in.

Reply:
It was a good idea to include the custom formatting function.

A simpler template format and finding replacement errors can be managed with text/template as here:
https://play.golang.org/p/QUJb5aA0IbF

Reply:
From a end user perspective, I like the simplicity of the `base-topic/{profilename}/device/{devicename}/state` syntax.

Reply:
Okay, yes it is simpler from the user's perspective. I'll update my implementation in the MQTT trigger to use Alex's method.